Real children (OC-88)

bradenmacdonald opened the pull request:
This removes LightChildren functionality from mentoring and converts all the LightChildren into full-fledged XBlocks.

I also moved all the former LightChildren into the "components" folder to make it easier to move them to separate repositories in the future.

I was also able to remove the dependence on Django models, though it still uses Django for template rendering and testing.

**Notes and questions**:
- XBlock names are very likely to conflict with other people's XBlocks... (e.g. `html` !). But the name of the block is determined by the tag name. May want to add a migration path to convert the XML to use namespaced blocks like `<mentoring-html>` instead of `<html>`. Or just make our HTML block the standard one and move it to a separate repository.
- Should not be installed on any system that actively uses other XBlocks depending on light_children (causes "ImportError: from mentoring import light_children")
- I got all the tests to pass but I think this needs more testing to be sure it won't break anything when installed onto edX instances already using the prior version
- What do I put as the copyright in new files, if anything? The current files all say "Copyright (C) 2014 Harvard"

bradenmacdonald committed:
Replace LightChildren w/ XBlock Children - break all the things

bradenmacdonald committed:
JS changes to get "Mentoring With Only One Step" scenario working

bradenmacdonald committed:
Changes to get mentoring assessments working.

bradenmacdonald committed:
Get some of the tests passing

bradenmacdonald committed:
Clean up some of the components

bradenmacdonald committed:
Fix sharing student responses among answers

bradenmacdonald committed:
Fix more tests

e-kolpakov commented on README.md:
````diff
@@ -481,29 +481,11 @@ Install to the workbench's virtualenv by running the following command form the
-root:
-
-```bash
-$ ./manage.py syncdb --settings=workbench.settings_mentoring
````
@bradenmacdonald correct me if I'm wrong, but isn't this step still required?

bradenmacdonald replied:
@e-kolpakov No, it's no longer required because I removed all the Django models from this repository, so it no longer needs to be added to `INSTALLED_APPS`. Instead, it uses XBlock setuptools entry points to register with workbench. So all that matters is that your installation of xblock-sdk is working and you have installed this xblock into the same virtual environment - it no longer needs any other configuration.

antoviaque replied:
@bradenmacdonald It would be nice to move to pure fields for answers, but there are then a few things to check:
- How do you maintain backward-compatibility with existing student data? Courses with existing answers will want to retain these answers. (This is btw a more general concern - existing data need to be retained for all fields that made use of persistence, ie things like assessment mode answers, max number of tries, etc.).
- Can data be accessed on XBlocks across different students? An important feature for Harvard is the ability to browse student answers from an admin dashboard - the initial naive implementation (put everything in a CSV) has been deactivated for performance reasons, but Harvard still wants it, and it will be back in the next set of changes as a dashboard allowing to browse the students answers. Would that be possible with pure XBlock fields?

bradenmacdonald replied:
@antoviaque 

Yes, after looking into this more, it's clear I will need to leave the Django models in place for now, at least in order to migrate the student data. (They can be kept separate and only need to be installed and used for migration purposes).

For accessing data across students I actually was just about to ask you about that. There are two options:
1. Keep using the Answers django model.
2. Convert to using standard XBlock storage as I have done so far - however, in order to do the data export that approach would require this one piece of code that uses non-XBlock APIs:

``` python
    def _get_students_data(self, answer_block):
        """
        Efficiently query for the answers entered by ALL students.
        (Note: The XBlock API only allows querying for the current
        student, so we have to use other APIs)

        Yields tuples of (student_id, student_answer)
        """
        usage_id = answer_block.scope_ids.usage_id
        # Check if we're in edX:
        try:
            from courseware.models import StudentModule
            usage_id = usage_id.for_branch(None).version_agnostic()
            entries = StudentModule.objects.filter(module_state_key=unicode(usage_id)).values('student_id', 'state')
            for entry in entries:
                state = json.loads(entry['state'])
                if 'student_input_raw' in state:
                    yield (entry['student_id'], state['student_input_raw'])
        except ImportError:
            pass
        # Check if we're in the XBlock SDK:
        try:
            from workbench.models import XBlockState
            for entry in XBlockState.objects.filter(scope="usage", scope_id=usage_id).exclude(user_id=None).values('user_id', 'state'):
                state = json.loads(entry['state'])
                if 'student_input_raw' in state:
                    yield (entry['user_id'], state['student_input_raw'])
        except ImportError:
            pass
        # Something else - return only the data
        # for the current user.
        yield (answer_block.scope_ids.user_id, answer_block.student_input_raw)
```

So the question is, which is worse - this code above, or requiring a separate Django model just so we can do data export?

antoviaque replied:
@bradenmacdonald Imho, as long as we have to go outside of the XBlock API, it can make sense to keep the current Django models approach - that likely reduces the amount of migration work and the potential unforseen pitfalls. It's still good to know that it's possible though, if we need to do this on existing XBlock fields.

Imho the real fix would be to add proper support for this - either within the XBlock API, via search, etc. Actually, it would be good to figure out what would be the right/canonical approach here, since this isn't the only change in the queue of upcoming work that will require accessing fields across different users (the other confirmed upcoming change will be to access answers in all mentoring MCQ fields from the current student and other students within a given cohort, to display the answers from everyone in the cohort in a single table for comparison). 

It would be worth posting about this on the ML imho, to see what Cale thinks is the right way to approach this (If you do, can you include the two use cases in the discussion?)

bradenmacdonald replied:
Ok. Will do.

I have posted on edx-code as well.

e-kolpakov commented on mentoring/components/table.py:
```diff
@@ -24,22 +24,22 @@
 
-log = logging.getLogger(__name__)
-
+loader = ResourceLoader(__name__)
```
@bradenmacdonald `loader` initialized twice

e-kolpakov commented on mentoring/dataexport.py:
```diff
@@ -30,18 +30,30 @@
+    writer = unicodecsv.writer(f, encoding='utf-8')
+    writer.writerow(row)
+    f.seek(0)
+    return f.read()
```
@bradenmacdonald you haven't closed `StringIO`. I'm not sure if it's smart enough to close itself when it's garbage-collected, but it would be better to do sometihng like

```
with StringIO() as f:
    writer = unicodecsv.writer(f, encoding='utf-8')
    writer.writerow(row)
    return f.getvalue()
```

bradenmacdonald replied:
I just moved this code unchanged from somewhere else, but that's a good point. Will fix :)

e-kolpakov commented on mentoring/mentoring.py:
```diff
@@ -36,42 +34,26 @@
+from components.title import TitleBlock
+from components.header import SharedHeaderBlock
+from components.message import MentoringMessageBlock
+from components.step import StepParentMixin, StepMixin
```
@bradenmacdonald I believe these all were imported into `components` namespace. Even if it's not it might make sense to do so, so that this block would become more semantical:

```
from components import TitleBlock, SharedHeaderBlock, MentoringMessageBlock, ...
```

e-kolpakov commented on mentoring/mentoring.py:
```diff
@@ -36,42 +34,26 @@
-    expected = _default_xml_content()
-
-    expected = re.sub(UUID_PATTERN, DUMMY_UUID, expected)
-    value = re.sub(UUID_PATTERN, DUMMY_UUID, value)
```
@bradenmacdonald this replaces are here for a reason - block may contain default content, but it contains url, which might contain uuids. I haven't got to the bottom of review yet, so I'm not sure if it's handled differently. Anyway, could you please confirm that new version works as expected?

bradenmacdonald replied:
Yep, I removed url_name from the template as the code to put it in no longer worked and it didn't seem necessary - see previous comment.

e-kolpakov commented on mentoring/mentoring.py:
```diff
@@ -83,54 +65,105 @@ class MentoringBlock(XBlockWithLightChildren, StepParentMixin):
-
-    display_name = String(help="Display name of the component", default="Mentoring XBlock",
-                          scope=Scope.settings)
+    xml_content = String(help="XML content", default=default_xml_content, scope=Scope.content)
```
@bradenmacdonald I'm not sure this exposes the same behavior as before. In previous version every XBlock got assigned unique `url_name` out of the box. As of now, they all get the same "malformed" url_name of `mentioring-`. This affects initial configuration only, but it's still changes existing behavior, and this change was not explicitly required. So, @antoviaque could you please have a look on it and approve/reject such behavior change?

bradenmacdonald replied:
@e-kolpakov This change was necessary because when I was running tests, I got test failures due to the fact that XBlock field defaults were not allowed to be callable in some cases. I'm not sure exactly which part of my system changed that requirement but I suspect it is because we need to target upstream edx-platform now instead of the solutions fork?

In any case, I assumed this was fine because in ALL of the example scenarios installed into workbench, the url_name functionality is not even used at all. It is only used in some of the tests, and those tests still pass. So whoever needs a `url_name` can still add it manually to the XML and use it. All that this changes is that there is no longer a default `url_name="random UUID"` showing up, which I suspect was something that people would always be changing anyways to use a human-readable name instead of `url_name`. But definitely let me know if my assumptions are wrong, either of you.

antoviaque replied:
@bradenmacdonald @e-kolpakov It's hard to be sure that we aren't forgetting edge cases with these changes, but greping the code I could only see this use having an impact:

https://github.com/edx-solutions/xblock-mentoring/blob/a8838b80f65a4d2066f523a31bdca69f05c89ad5/mentoring/mentoring.py#L193

With a non-random url_name it means the component_id logged in events would be the same for all newly created mentoring XBlocks for which this isn't edited. We might be able to get away with using another unique id there if the url_name is the default one when an event is emitted.

But we should really be using the feature we specifically contributed for this use case: https://github.com/edx/XBlock/pull/249 If the XBlock repo is out of sync with edx-platform, if might be worth pushing a PR on edx-platform for this, and ask if this can get into birch.

bradenmacdonald replied:
@antoviaque I changed the code to use the XBlock's usage_id if the url_name has not been set.

The new feature would be nicer though - is there still time to get things into Birch?

antoviaque replied:
@bradenmacdonald That's what I suggest to find out by pushing a requirements update PR and asking about it - I would ping Cale & DB to ask them, pointing at the feature we'd like to see included in birch.

bradenmacdonald replied:
@antoviaque DB says it won't make it into Birch (PR 6791). Will our clients generally be sticking to the named releases? If so I may need to leave this as-is for now and not use the new UNIQUE_ID feature...

Kelketek replied:
@bradenmacdonald @antoviaque I've been seeing this thorn in our side keep coming up. Do we even need to bother with `url_name`? That's automatically generated at random by the CMS. Why do we handle it manually here? If it's omitted from the XML, it will still be set under the hood. Is there something special we're doing with it?

bradenmacdonald replied:
@Kelketek Interesting - I didn't actually realize that `url_name` was a thing in the CMS code. The problem is that `url_name` is not part of the XBlock spec and doesn't appear anywhere in the XBlock code or the SDK - so if we want workbench compatibility we can't rely on that behaviour and have to implement it ourselves. Please correct me if I'm wrong...

For any new XBlocks, we should really just be using `usage_id`s and the `Reference` field type for inter-block identification and references, as I believe that's the [emerging] standard.

Kelketek replied:
@bradenmacdonald Yes, that is correct. You may have some issue with it on the workbench, but we don't need so robust an approach as we're using right now for that case. You may just be able to specify the url_name in the XML for the test scenarios. If that's not good enough, just make sure that some bogus value is handed around that is never actually saved should the url_name not be available and be needed, and never bother with saving it. That should be good enough for testing.

Kelketek replied:
If you do manage to eliminate our weirdness with url_name this way, you may allow us to start using a much more recent XBlock version. Is there anything, other than `FIELDS_TO_INIT` that we're dependent on our fork for?

antoviaque replied:
@bradenmacdonald Yup, I saw - too bad for birch :/ And yep, at least McKinsey will stick to the named releases, so we would have to keep it as is until the next release.

@Kelketek This is already using a newer version of the XBlock repo, the real children refactor depends on the current edx-platform master. When we rebase the edx-solutions fork, we will rebase the xblock repo fork too (along with all repos). I'm not sure what moving away from url_name would gain us here?

Kelketek replied:
@antoviaque Well, for one, it would remove us having to fight with url_name every time we do something to this repo. It's happened on almost every task we've done with Mentoring since I got here. If we can side-step the fight altogether, that would be my preference. If we don't do it in this PR, we should make a task of it, because it will save us time down the line. If we find we still need that functionality, then when the feature is available to us upstream, we can take advantage of it. But either way I suspect we can greatly simplify how we're handling it here.

I wasn't aware this was using the upstream XBlock, but I'm curious: What do we currently have in our XBlock fork that isn't implemented upstream? I know we've added our own implementation of handling children, and FIELDS_TO_INIT, which is to be replaced with a different solution. I'm mostly curious if there's some reason that our rebase of XBlock can't be a git reset --hard against birch.

antoviaque replied:
@Kelketek Unless I've missed something, the reason there has been multiple changes around url_name is because we have been struggling to find the right approach for initializing the field that would get the assent of upstream - there has been multiple iterations over this. But now that we have a good and agreed solution merged, and that we only have to wait for it to become available, I'm not sure if we want to go yet another way, with potentially a new set of issues - the workbench compatibility you were discussing being one example.

For the XBlock fork, you can see the difference here: https://github.com/edx-solutions/XBlock/commits/master there are only the FIELDS_TO_INIT and your reimplementation of child serialization, so we should likely be able to take the current XBlock master (not the one pointed by birch though, which is pointing to an older XBlock commit currently).

Kelketek replied:
@antoviaque Yes, I suppose that's true, on why we've been fighting with it. I guess my question is why we had to start fighting with it to begin with. I only relatively recently learned what url_name actually was when making the Poll XBlock. It just made me wonder if all the trouble we went through to get an upstream solution made was necessary, or if it came out of a general misunderstanding of what it was.

I suppose it doesn't too much matter now-- it's good to have the feature upstream either way, since now we have a standardized way to handle it. Sounds good to me. :)

e-kolpakov commented on mentoring/mentoring.py:
```diff
@@ -83,54 +65,105 @@ class MentoringBlock(XBlockWithLightChildren, StepParentMixin):
-        """Compute the student score taking into account the light child weight."""
-        total_child_weight = sum(float(step.weight) for step in self.steps)
+        """Compute the student score taking into account the weight of each step."""
+        weights = [float(self.runtime.get_block(step_id).weight) for step_id in self.steps]
```
@bradenmacdonald nit: could have used generator instead of list comprehension. Might save us a bit of headache if someone have 10000+ step blocks :)

bradenmacdonald replied:
@e-kolpakov OK I can change, but a list of 10,000 floats should be no problem anyways, right? A list of 10,000 XBlocks would be a problem...

e-kolpakov commented on mentoring/mentoring.py:
```diff
@@ -144,14 +177,31 @@ def student_view(self, context):
+        child_content = u""
+        for child_id in self.children:
+            child = self.runtime.get_block(child_id)
+            if isinstance(child, TitleBlock):
```
@bradenmacdonald oh, you redecorated it... I don't like it. :) Being serious - I don't like this loop. I understand what's happening here, and it should just work, but it's a bit clumsy. Can't we first separate special-case blocks into their own list(s), than only process usual ones like this, and give special treatment to specials? Something like

```
title_blocks, header_blocks, messages_blocks, ordinary_blocks = self._split_children()
# _split_children is probably going to do self.runtime.get_block(child_id) internally, than return actual blocks
for child_block in ordinary_blocks:
   ....
title = title_blocks[-1].content
header = header_blocks[-1].render(...)
```

Note I'm using `-1` to get last block just to preserve the behavior this code exhibits now - later blocks overwrite former ones. It might be more than that, e.g. asserting that there are only one such block, etc.

bradenmacdonald replied:
@e-kolpakov I've simplified the code a bit, but I believe this approach is more efficient than something like self._split_children() because this way only one child block is ever loaded at a time. If we split the sorting logic out into another method, then it will have to load all the children at once (unless I make the code significantly more complicated).

e-kolpakov commented on tests/integration/xml/mcq_with_comments_1.xml:
```diff
@@ -12,8 +12,6 @@
             <tip require="beauty" width ="400" height = "600">This is something everyone has to like about beauty</tip>
             <tip reject="bugs" width = "100" height = "200">Nah, there isn\'t any!</tip>
-
-            <!--<message type="on-submit">Thank you for answering!</message> -->
```
@bradenmacdonald this comment was intentional - there were bugs with mentoring containing XML comments. This is a XML fixture to be used in test that verified comments are just ignored. If tests fail than we probably need to update XBlock so that when it parse XML it ignore comments. Also, haven't you noticed the name of the file? :)

I think I understand now why have you removed it - it was failing all the integration tests - that's because workbench builds a scenario out of each XML stored in tests/xml, and one with comments just crashed every time it was touched. I believe the correct way to handle this would be to replicate the fix that this test was intended to test in XBlock repository (and probably upstream)

bradenmacdonald replied:
Oh, whoops! I didn't notice the name of that file. Will definitely fix :) If I recall, it will just require some weird code due to the conventions of python's XML library when handling comments.

antoviaque commented:
@bradenmacdonald 

> XBlock names are very likely to conflict with other people's XBlocks... (e.g. html !). But the name of the block is determined by the tag name. May want to add a migration path to convert the XML to use namespaced blocks like `<mentoring-html>` instead of `<html>`. Or just make our HTML block the standard one and move it to a separate repository.

Yes, ideally we would make ours the standard one, or eventually migrate to use the standard `<html>` XBlock that emerges. The main point being that we should try to aim toward reusing standard pieces rather than reinventing the wheel on each XBlock. Plus, we'll need to keep the syntax compatible with existing courses using `<mentoring>` (edx-solution and Harvard).

> Should not be installed on any system that actively uses other XBlocks depending on light_children (causes "ImportError: from mentoring import light_children")

I see - you are referring to the Adventure XBlock, right? It will need to be converted too, yes.

> I got all the tests to pass but I think this needs more testing to be sure it won't break anything when installed onto edX instances already using the prior version

Yes, it's important. I would suggest exporting courses from integration or QA, like the Demo course or MAPS, and testing them in the LMS and apros.

> What do I put as the copyright in new files, if anything? The current files all say "Copyright (C) 2014 Harvard"

There are multiple copyrights at that point - I would put:

Copyright (c) 2014-2015 Harvard, edX & OpenCraft

bradenmacdonald committed:
Cleanups

bradenmacdonald committed:
Add Travis CI

bradenmacdonald committed:
PEP8 Test and fixes, add three pylint checks as well

bradenmacdonald committed:
Add build status icon

bradenmacdonald committed:
Fix+test data export, fix multiple <mentoring> blocks on one page

bradenmacdonald commented:
@e-kolpakov @antoviaque Update on this: I have addressed all comments so far, and added a new test for the CSV export. Eugeny, if you want to review again now that would be great, but if you want to wait until everything is done, no problem.

I have also set up Travis CI on this repository to run the tests and a few basic quality checks. As you can see, it's working quite nicely :) If you click on the build status link you can even see a coverage report.

Still to do:
- Decide if we prefer to keep storing Answer data in Django model or not.
- Add migration code to migrate old data (from light children and from answer django tables)
- Test with actual course data to make sure it works, test more extensively with edx-platform and not the workbench
- Re-implement the `jump_to_id` linking code that I saw in here but which wasn't covered by any tests or examples AFAIK.
- Improve coverage

e-kolpakov commented on mentoring/components/questionnaire.py:
```diff
@@ -56,34 +51,36 @@ class QuestionnaireAbstractBlock(LightChild, StepMixin):
-
-        for name, value in attr:
-            setattr(block, name, value)
+            elif xml_child.tag is not etree.Comment:
```
@bradenmacdonald this is not exactly a correct way to fix this. In previous version it was fixed like this: https://github.com/open-craft/xblock-mentoring/commit/1a6e2092a361f0befff18d07607517d437046e3b#diff-23a9d1f1d9cda06ac8fa67c8c2f4589dR130-131

Now the code that deserializes XML into mentoring XBlocks are in main XBlock repository: https://github.com/edx/XBlock/blob/9c634481dfc85a17dcb3351ca232d7098a38e10e/xblock/runtime.py#L685

This is probably a good idea to teach upstream XBlock to play nicely with XML comments

bradenmacdonald replied:
@e-kolpakov Why is this not correct? This is from the lxml docs...

I agree it would be better to put the fix upstream, though for now we need to include a fix locally.

e-kolpakov replied:
@bradenmacdonald mostly because when I did something similar I ended up going with comment-ignoring parser. I'm sorry for using "incorrect" (not actually using, but it sounded like that), I meant more like "this would work, but there's potentially better way". Please forgive me, I'm not native english-speaker :)

bradenmacdonald replied:
@e-kolpakov  Ah, no problem. I took your advice and have opened PR 288 against upstream XBlock to fix this there. For now I'm leaving the workaround code in this repository though.

e-kolpakov commented on mentoring/dataexport.py:
```diff
@@ -23,25 +23,42 @@
+    """
+    Convert a list to a CSV string (single row)
+    """
+    f = StringIO()
```
@bradenmacdonald I might repeat myself, but it's still better written using context manager.

bradenmacdonald replied:
@e-kolpakov I tried a context manager but it turns out that `StringIO` doesn't support that.

e-kolpakov replied:
@bradenmacdonald awww, it was added in python3. It happens to me sometimes :) Suggesting features from other versions/languages :)

e-kolpakov commented:
:+1: conditional @antoviaque and upstream reviewers approve `xml_child.tag is not etree.Comment` instead of comment-ignoring XML parser

bradenmacdonald committed:
Consolidate some code and speed up StepParentMixin

antoviaque commented:
@e-kolpakov Yup, the alternate fix LGTM - I think in the discussion on the original fix my concerns were more about the fact that https://github.com/open-craft/xblock-mentoring/commit/1a6e2092a361f0befff18d07607517d437046e3b#diff-23a9d1f1d9cda06ac8fa67c8c2f4589dL102 was too generic (ie not targetting comments specifically - see https://github.com/edx-solutions/xblock-mentoring/pull/74#discussion_r17160344 ).

e-kolpakov commented:
@antoviaque @bradenmacdonald any chance to see it merged any time soon?

antoviaque commented:
@e-kolpakov The ticket wasn't in ready for merge so I asked confirmation from @bradenmacdonald on the Jira ticket. But if all is good for you and you need the merge, I would say go ahead and merge - another PR can always be reopened if necessary.

bradenmacdonald commented:
@e-kolpakov This branch is not yet backwards-compatible with older versions. I'd like to reinstate the Answers Django model and test for backwards compatibility before we merge.

bradenmacdonald committed:
Restore Django models - preserve Answer data

bradenmacdonald commented:
@e-kolpakov @antoviaque OK, I think this is ready to merge after one of you has looked at my last commit - which is mostly undoing some of my other changes. I have now reinstated the `Answer` model so that we can preserve student answers and produce the instructor reports without circumventing the XBlock API.

Next sprint we can test against real course data and make sure it migrates properly.

e-kolpakov commented:
@bradenmacdonald I see no criminal in last commit :+1:

antoviaque commented:
@bradenmacdonald @e-kolpakov Thanks! Btw, when you have the rights on a repository and you have got the necessary :+1: from your reviewer, feel free to merge directly without waiting for me. Still make sure to ping me on the PR to ensure that I can update clients as necessary, but my added value in clicking that "Merge" button is very small : )
